### PR TITLE
Adding docdb-scaleOnSchedule.py to Lambda samples

### PR DESCRIPTION
**Pull request by kaarthiik-aws**
*Issue #, if available:*

*Description of changes:*
Added docdb-scaleOnSchedule.py. Using this Lambda function, users can add or remove instances from their cluster. Users have to create two Eventbridge rules, one to schedule the addition and the other to delete instances.
Updated README to add a section for docdb-scaleOnSchedule.py

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                Engine = 'docdb',
+                PromotionTier = instancePromotionTier )
+                
+                createResponse[instanceIdentifier] = instanceCreateResponse['ResponseMetadata']
+                
+            except:
+                logger.error("Encountered problem while creating instance. Error message is ", sys.exc_info()[0], sys.exc_info()[1])
+             
+    else:
+            logger.info( "Cannot add more instances to the cluster " + clusterIdentifier + " already has  " + str(currInstanceCount) + " instances. " + "Adding " + str(numInstancesToAdd) + " instaces exeeds the maximum number of allowed instances (16) in a cluster")
```
Comment (mihaialdoiu) [inline lambda-samples/samples/docdb-scaleOnSchedule.py:62]: A typo here: "instaces exeeds"

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
+import json
+import os
+import boto3
+import random
+import logging
+import string
```
Comment (mihaialdoiu) [inline lambda-samples/samples/docdb-scaleOnSchedule.py:6]: Do you need string? I don't see it being used.

File: lambda-samples/samples/README.md [hunk 3]
```diff
@@ -70,3 +70,43 @@ Python 3.x
...
+                "rds:ListTagsForResource",
+                "rds:AddTagsToResource",
+                "rds:DescribeDBClusters"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+- Configure the lambda timeout to a higher value than the default 3 seconds.
```
Comment (mihaialdoiu) [inline lambda-samples/samples/README.md:103]: How much is recommended?

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import logger
+import sys
+
+#Set up the logger
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+clusterIdentifier = os.environ['CLUSTER_IDENTIFIER']
+instanceSuffix = os.environ['INSTANCE_NAME_SUFFIX']
```
Comment (mihaialdoiu) [inline lambda-samples/samples/docdb-scaleOnSchedule.py:16]: Add validation, as instance names must begin with a letter; must contain only ASCII letters, digits, and hyphens; and must not end with a hyphen or contain two consecutive hyphens.

**Review by mihaialdoiu — COMMENTED** [review]
Works, I would clean it up a bit, make sure to import only what's needed, add comments for each function, check for spelling errors. See my specific comments.

File: lambda-samples/samples/docdb-scaleOnSchedule.py (hunk 2 above)
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py]: It is used in line 72 : randString = "".join(random.choices(string.ascii_lowercase, k=6))

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+import json
+import os
+import boto3
+import random
+import logging
+import string
+import logger
```
[the remark this answers is not in the document]
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py]: Removed

File: lambda-samples/samples/docdb-scaleOnSchedule.py (hunk 4 above)
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py]: Updated script to used a fixed suffix, instance name suffix is not taken from user input

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import sys
+
+#Set up the logger
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+clusterIdentifier = os.environ['CLUSTER_IDENTIFIER']
+instanceSuffix = os.environ['INSTANCE_NAME_SUFFIX']
+instanceClass = os.environ['INSTANCE_CLASS']
```
[the remark this answers is not in the document]
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py]: Intended to provide user flexibility on choosing different instance type for read scaling. And, there is a possibility cluster may have heterogeneous instances.

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+instancePromotionTier = int(os.environ['INSTANCE_PROMOTION_TIER'])
```
[the remark this answers is not in the document]
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py]: Thank you for the suggestion. Updated script, promotion tier is no more an user input.

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+numInstancesToAdd = int(os.environ['INSTANCES_TO_ADD'])
+numInstancesToDel = int(os.environ['INSTANCES_TO_DELETE'])
+
+client = boto3.client('docdb')
+
+def lambda_handler(event, context):
```
[the remark this answers is not in the document]
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py:45]: Added validation for instance class.

File: lambda-samples/samples/docdb-scaleOnSchedule.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                DBInstanceIdentifier = instanceIdentifier,
+                DBInstanceClass = instanceClass,
+                Tags = [{'Key':'scheduledTemporaryInstance', 'Value':'Yes'}],
+                Engine = 'docdb',
+                PromotionTier = instancePromotionTier )
+                
+                createResponse[instanceIdentifier] = instanceCreateResponse['ResponseMetadata']
+                
+            except:
+                logger.error("Encountered problem while creating instance. Error message is ", sys.exc_info()[0], sys.exc_info()[1])
```
[the remark this answers is not in the document]
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py]: My bad, missed it. Now corrected

File: lambda-samples/samples/docdb-scaleOnSchedule.py (hunk 1 above)
Comment (kaarthiik-aws) [reply lambda-samples/samples/docdb-scaleOnSchedule.py]: Corrected

**Review by kaarthiik-aws — COMMENTED** [review]
Implemented suggested changes.

**Review by tmcallaghan — APPROVED** [review]
Approved.

**Pull request merged**